tests/test_product.py

```diff
@@ -0,0 +1,64 @@
+import pytest
+from buy_together_app import models
+# from buy_together_app import views
+
+QR = "Q5o76MbdiNbXprNEnHfpcGWFp1CMF8XY"
+NAME = "Banana"
+DESCRIPTION = "Good!"
+
+
+@pytest.fixture
+def product0():
+    return models.Product(qr_code=QR, product_name=NAME, description=DESCRIPTION)
+
+
+class TestProductModel:
+    def test_new_product(self, product0):
+        assert product0.qr_code == QR
+        assert product0.product_name == NAME
+        assert product0.description == DESCRIPTION
+
+    @pytest.mark.django_db()
+    def test_save_product(self, product0):
+        product0.save()
+        # print(models.Product.objects.all()[0].product_name)
+        # assert False
+        assert product0 in models.Product.objects.all()
+
+    @pytest.mark.django_db()
+    def test_delete_product(self, product0):
+        product0.save()
+        # print(models.Product.objects.all())
+        # assert False
+        product0.delete()
+        assert product0 not in models.Product.objects.all()
+
+    # @pytest.mark.skip
+    @pytest.mark.django_db()
+    def test_filter_by_qr(self, product0):
+        product0.save()
+        # print(list(models.Product.filter_qr(product0.qr_code)))
+        # assert False
+        assert [product0] == list(models.Product.filter_qr(product0.qr_code))
```

Comment:
I think you should add additional products to the db. Otherwise, you can have false positives here for the cases in which you return all products.

Reply:
It will also be good to have a test case for those cases where you expect to get back all products.

Reply:
olny if qr is none i get all products and assert will prevent that,other wise i will get the same fixture in both sides
i do have testcase for all products

Reply:
I meant to test for the `qr is None` case. It's in your code so it should have test coverage.

Reply:
Done

Reply:
Consider this is the method you're testing:
```
    @staticmethod
    def filter_qr(qr):
        if qr is not None:
            return Product.objects.filter(qr_code=qr)
        return Product.objects.all()
```
A good test coverage means that you're testing all your flow of control branches. i.e. all sides of if, for, while statements. In this case, you have a function that have two return statements. You need to have test cases for both of them.

Namely, one in which the qr is specified and you return a product, and another in which it's not and then you return all products.

Reply:
(and this goes as well for the other filters)

Reply:
pytest provides a way to generate multiple use cases in one test using test parameterization. I don't want you to waste too much of your time on making that work now, so you don't have to use it now, but just have it in mind for the future: https://docs.pytest.org/en/6.2.x/parametrize.html

Reply:
Done